### ISSUE TITLE
Rule to validate correct @ts-ignore usage

### ISSUE THREAD

**Issue by Swatinem** [palantir/tslint#3975]
### Rule Request

I would like to propose a rule that does at least the following:

#### forces `@ts-ignore` to be accompanied by a comment

The reasoning here is that you should not blindly throw `@ts-ignore` comments around yourself, but explain your reasoning to do that, even better, point to an issue for further explanation

**valid**
```ts
// @ts-ignore https://github.com/Microsoft/TypeScript/issues/11152
const Apps = new Map(apps.map(app => [app.name, app]))
```
**invalid**
```ts
// @ts-ignore
const Apps = new Map(apps.map(app => [app.name, app]))
```

#### error on `@ts-ignore` annotations that do not actually suppress any error

**valid**
```ts
// @ts-ignore
const foo: string = 123;
```
**invalid**
```ts
// @ts-ignore
const foo: number = 123;
```

**Comment by suchanlee** [comment]
I think that this is a legitimate use-case and can be a useful rule to enforce via lint

**Comment by giladgray** [comment]
is it even possible to practically (efficiently) detect when `@ts-ignore` is applied to TS code that _does not produce an error_?

**Comment by Swatinem** [comment]
> is it even possible to practically (efficiently) detect when @ts-ignore is applied to TS code that does not produce an error?

I’m not sure about that. Maybe changes in the TS typechecker itself are needed.
See the linked https://github.com/Microsoft/TypeScript/issues/25166 for that, and also a very good explanation for the reasoning why this should be useful.

Flow also has the behavior to error on ineffective suppression comments.

**Comment by giladgray** [comment]
ah nice an issue on the language itself! I would argue very strongly that this feature belongs _in `tsc` itself_ and not in tslint because it's a compiler flag and not a style decision. i'm going to close this issue to redirect the conversation to that TS issue.

**Comment by Swatinem** [comment]
I agree that warning on ineffective suppression should be a `tsc` feature.

But the second part of this feature request is actually a linting issue:

* making sure that `@ts-ignore` comments are accompanied by some kind of explanation why we need to suppress errors there.

So I would appreciate it if you would reopen this for the sake of the other feature I requested :-)

**Comment by giladgray** [comment]
please file a new issue on the request above. this one is officially hosed (due to the link in the TS issue)

**Comment by dbkaplun** [comment]
This ticket shouldn't have been closed. The TypeScript team has not responded yet, so it is unclear that the solution will actually be implemented in TS itself.
